src/h3c_fmt_plug.c

```diff
@@ -0,0 +1,320 @@
+ * modification, are permitted.
+ *
+ * There's ABSOLUTELY NO WARRANTY, express or implied.
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_h3c;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_h3c);
+#else
+
+#include <string.h>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#include "arch.h"
+#include "common.h"
+#include "formats.h"
+#include "params.h"
+#include "options.h"
+#include "johnswap.h"
+#include "base64_convert.h"
+#include "sha2.h"
+#include "simd-intrinsics.h"
+
+#define FORMAT_LABEL       "h3c"
+#define FORMAT_NAME        "H3C/Huawei/HPE"
+#define FORMAT_TAG         "$h$6$"
+#define TAG_LENGTH         (sizeof(FORMAT_TAG)-1)
+
+#ifdef SIMD_COEF_64
+#define ALGORITHM_NAME     SHA512_ALGORITHM_NAME
+#define NBKEYS                     (SIMD_COEF_64*SIMD_PARA_SHA512)
+#else
+#define ALGORITHM_NAME     "64/" ARCH_BITS_STR
+#define NBKEYS             1
+#endif
+
+#define BENCHMARK_COMMENT  ""
+#define BENCHMARK_LENGTH   0x107
+#ifdef SIMD_COEF_64
+#define PLAINTEXT_LENGTH   ((128-16-8-1)/2-1)
+#else
+#define PLAINTEXT_LENGTH   125
+#endif
+#define BINARY_SIZE        8
+#define BINARY_ALIGN       4
+#define SALT_SIZE          16
+#define SALT_ALIGN         1
+#define MIN_KEYS_PER_CRYPT NBKEYS
+#define MAX_KEYS_PER_CRYPT (64*NBKEYS)
+#define DIGEST_SIZE        64
+#define CIPHERTEXT_LENGTH  88
+
+#define BASE64_ALPHABET    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="
+
+#ifndef OMP_SCALE
+#define OMP_SCALE          8
+#endif
+
+#ifdef SIMD_COEF_64
+#define FMT_IS_64BIT
+#endif
+
+static unsigned char cursalt[SALT_SIZE];
+static int (*saved_len);
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static uint64_t(*crypt_out)[DIGEST_SIZE / sizeof(uint64_t)];
+
+
+static struct fmt_tests h3c_tests[] = {
+	{
+		FORMAT_TAG "4tWqOiqovcWddOKv$XyFMVgaE46fGiqsZEHbcr+BM/m9tDkvahDbqU7HoNrvmALk2u31z9c/tuUmX7IiQhWRwN5qoZquW82A8XYaDWA==",
+		"abc"
+	},
+	{
+		FORMAT_TAG "abcdefghijklmnop$jp3hDbVlf/L1GNDE4n6x4wqvHnFiEr4YrtM6ax1aFXFb6pdu4Nfmpp09pZFOGOH8ID9vw2AOKQA4q8lByhlG4A==",
+		"password"
+	},
+	{ NULL }
+};
+
+static void init(struct fmt_main *self)
+{
+	omp_autotune(self, OMP_SCALE);
+	saved_len = mem_calloc(self->params.max_keys_per_crypt, sizeof(*saved_len));
+	saved_key = mem_calloc(self->params.max_keys_per_crypt, sizeof(*saved_key));
+	crypt_out = mem_calloc_align(self->params.max_keys_per_crypt, sizeof(*crypt_out), sizeof(uint64_t));
+}
+
+static void done(void)
+{
+	MEM_FREE(crypt_out);
+	MEM_FREE(saved_key);
+	MEM_FREE(saved_len);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	int len;
+
+	if (strlen(ciphertext) != TAG_LENGTH + SALT_SIZE + 1 + CIPHERTEXT_LENGTH)
+		return 0;
+
+	if (strncmp(ciphertext, FORMAT_TAG, TAG_LENGTH))
+		return 0;
+
+	ciphertext += TAG_LENGTH;
+
+	if (ciphertext[SALT_SIZE] != '$')
+		return 0;
+
+	ciphertext += SALT_SIZE + 1;
+
+	len = strspn(ciphertext, BASE64_ALPHABET);
+	if (len != CIPHERTEXT_LENGTH)
+		return 0;
+	return 1;
```

Comment:
I see we have similar logic in `sha512_common_valid_nsldap`, and in both places we don't fully check that the base64 encoding is valid (`=` only used as padding). We might want to enhance both with a test `base64_convert` run and checking its return value. A way to do that may be to have `get_binary` return NULL if `base64_convert` fails to provide the expected length data, call `get_binary` from `valid` and check its return value. Since we would also be fixing the pre-existing code, this (or at least that part) should be a separate commit (may also be separate PR).